Pull-up with recent master

MarcelHB opened the pull request:
Now I noticed the full degree of disruption by my recent changes.

For the first step, I merged the current HEAD and hammered things around to make them work and compile green. But since `GemRB_Window_SetPicture` already disappeared and my IWD2 night-menu screen changes made use of that, the main menu is already a little broken. I guess I will have to restore that method. :)

kujeger committed:
use SDL_WINDOW_RESIZABLE and SDL_WINDOW_FULLSCREEN_DESKTOP
to avoid changing display mode. Let SDL2 handle resizing transparently.

kujeger committed:
Clear so toggling fullscreen is better. Also set scaling method.

kujeger committed:
ugly hack as workaround for cursor trails in SDL2

kujeger committed:
ah, indentation.. no spaces for me.

kujeger committed:
only copy backBuf is cursor is visible

fizzet committed:
fix RandomWalk logic with MaxWalkDistance set

I didn't previously realize RandomWalk used RunAway instead of FindPath...

lynxlynxlynx committed:
GetModState: added overloaded variant to shorten most uses

lynxlynxlynx committed:
SetupItemAbilities: don't bail out on missing tooltips

lynxlynxlynx committed:
GCW: differentiate between qitem and qweapon right-clicks

lynxlynxlynx committed:
GetItem: expose item extended header's ability locations

lynxlynxlynx committed:
SetupItemAbilities: show only the needed headers

lynxlynxlynx committed:
Merge pull request #39 from kujeger/sdl2-hacking

Sdl2 hacking (window resizing&scaling, cursor trail hack)

fizzet committed:
only clear actions when interrupts haven't been disabled

Fixes Albert not disappearing in rufcut01.

lynxlynxlynx committed:
Actor: randomise nextComment, so we can play it sooner without cacophonies

Commit:
operator precedence bugs: ! binds more tightly than &

lynxlynxlynx committed:
added detection and use of __attribute__ ((unused))

gcc5 is stricter ...

lynxlynxlynx committed:
SDL20VideoDriver::ProcessEvent: added another crash guard

lynxlynxlynx committed:
AbilitiesItemWindow: only respec the button in bg2

fixes the rest not showing the selection "border"

lynxlynxlynx committed:
SDL: init tmpBuf

coverity ...

lynxlynxlynx committed:
BMPImporter: init members

lynxlynxlynx committed:
BIFImporter: init members

lynxlynxlynx committed:
SetupItemAbilities: select the equipped ammo last

if the first quiver was equipped, it would never show, since its state got reset later
(suspect radio)

lynxlynxlynx committed:
SetupItemAbilities: use a frame with a border for the selected state

lynxlynxlynx committed:
SetupItemAbilities: fixed qitem disabled frame

it uses a different select frame, so the numbers don't match the previous commit

kujeger committed:
Fix "reverse" scroll in SDL2 when too tall or wide

when SDL2 scales the window to a different aspect ratio, it will pad
sides with black.
Since mouse cursor position was passed unsigned, the negative x and/or y
values generated when in the left and/or top padded areas was wrapped,
and scrolling went the wrong way.

bradallred committed:
Merge pull request #40 from kujeger/sdl2-hacking

Fix "reverse" scroll in SDL2 when too tall or wide

lynxlynxlynx committed:
trivial comment change so nano's syntax highlighter doesn't break

lynxlynxlynx committed:
Bitmap: added dumper function

lynxlynxlynx committed:
OpenGLEnv: fixed the include header case for gles

lynxlynxlynx committed:
cmake: check for gles before committing to it

lynxlynxlynx committed:
fx_set_slowed_state: also decrease spellcasting speed

kujeger committed:
set blend mode in SetSurfaceAlpha

SDL_SetSurfaceBlendMode needs to be SDL_BLENDMODE_BLEND on the surface
in order for it to, well, blend.

lynxlynxlynx committed:
Merge pull request #41 from kujeger/sdl2-hacking

set blend mode in SetSurfaceAlpha

lynxlynxlynx committed:
travis: try fixing the pull request tester

it doesn't pull the secrets from travis, so all the encrypted vars are empty;
however we only need them for deployment, which doesn't run here

fizzet committed:
print correct armor class in levelup window

lynxlynxlynx committed:
cmake: added raspberry pie detection

lynxlynxlynx committed:
SDLAudio::Init: added workaround for lagging with rpy

may become redundant once more of the plugin is implemented and the buffer fills faster

fizzet committed:
BG1/2: disable walking sounds for bats, rats, squirrels...

fizzet committed:
BeginDialog: fix protagonist checks

lynxlynxlynx committed:
Actor: added GetAdjustedTime for AdjustedTicks related time dilation purposes

lynxlynxlynx committed:
fixed regeneration and poison to take adjusted time into consideration

lynxlynxlynx committed:
fx_cast_spell_on_condition: account for adjusted time

doesn't make much sense, but that's what the original did

lynxlynxlynx committed:
SlicedStream::Write: converted to an assert

never properly implemented (inf loop), which now also fixes a clang build problem

lynxlynxlynx committed:
Actor: modal actions also account for adjusted time

the only remaining thing besides indirectly combat is animation speed:
slowed actors also have their animation speed halved

lynxlynxlynx committed:
NEWS: centirev bump

lynxlynxlynx committed:
travis: a different attempt at fixing pull request test builds

lynxlynxlynx committed:
update the tiles on day-night transitions

fixes Edwin's Docks (ar0300) entrance not updating all its (new) tiles,
leaving lightmap artifacts - tiles from the previous wed

lynxlynxlynx committed:
CircularBuffer: don't crash if there's nothing to retrieve

lynxlynxlynx committed:
GUIScript::ExecString: don't crash if the string has spaces (TypeError)

lynxlynxlynx committed:
trivial space cleanup to iwd2 itemuse.2da

MarcelHB committed:
IW2::CG: adjusts skill points calculation

When creating a new PC, there were more skill points available than
there were supposed to be.

This changes the partial formula from:

(class pts + human pts + int bonus pts) * 4

to:

(class pts + human pts) * 2 + int bonus pts * 4

This returns values equal to the original game.

lynxlynxlynx committed:
Merge pull request #43 from MarcelHB/iwd2_cg_skill_points

IWD2::CG: adjusts skill points calculation

lynxlynxlynx committed:
iwd2: check class item usability before kit usability

kujeger committed:
Sdl2 input hacking (#42)

SDL2: fixed several keyboard input problems

* SDL2: also skip scancodes 32 and 1073742053

scancode 32 is space, 1073742053 is shift -- not "skipping" these means that various text input fields put space twice, and also "printed" the shift keypress

* SDL2: "Skip" shift+ctrl. Fix key-combos.

Re-implement the fullscreen and console keycombo in the SDL2 code,
and "skip" not just one shift button, but both of them plus ctrl.

* Move key ignoring to SDL_KEYDOWN, alter clauses.

SDL_KEYUP events are now passed through to SDLVideoDriver::ProcessEvent,
since the SDL_TEXTINPUT handler only passes keys to
EvntManager->KeyPress -- that is, only as if they were pressed. This
meant that keys snagged by SDL_TEXTINPUT could only ever generate events
as if KEYDOWN, never KEYUP.
Keys are not also passed on unfiltered if they are pressed while holding
ctrl, to allow key-combinations like fullscreen and console

* SDL2: Handle console popping directly. Fix numpad.

Since console is the only time a KEYDOWN shortcut is handled directly in
SDLVideoDriver::ProcessEvent, it is reproduced in the SDL2 code.
Also add keypad buttons to "filter" so they work as in SDL1.

MarcelHB committed:
IWD2::Skills: prevents scrolling beyond skill items

In addition, this saves a Python exception.

MarcelHB committed:
GUI::EventMgr: preserves focus lock after rk interrupt

If `rk_flags' caused a GEM_MB_ONGOING_ACTION, we keep the focus anyway.
Otherwise, we couldn't use a scollbar outside of its containment.

MarcelHB committed:
GUI::ScrollBar: preserves active state on ongoing action

Scrollbar scrolling no longer accidentally finishes on leaving the
containment and re-entering while grabbing the handle.

lynxlynxlynx committed:
iwd2: fixed skipping of racial enemy

lynxlynxlynx committed:
Merge pull request #44 from MarcelHB/iwd2_skill_window

IWD2::Skills: prevents scrolling beyond skill items

MarcelHB committed:
String: extends buffer in `MBCStringFromString'

If `String' aliases to `std::wstring', `::length' returns the number
of wide-chars, not bytes. For the worst case, we should allow each
`wchar_t' to be mapped into space not smaller, shrinking the space after
the result of `wcstombs'.

Otherwise, multibyte charachters make the resulting string cut-off
early.

lynxlynxlynx committed:
iwd2: fixed stat resetting in OpenSkillsWindow

alternative to PR #48

lynxlynxlynx committed:
iwd2: completed fix from 0f2d12fbc305 to work properly for other races

lynxlynxlynx committed:
Merge pull request #46 from MarcelHB/mbstring_fix

String: extends buffer in `MBCStringFromString'

MarcelHB committed:
SDLVideo/GUI: changes `MouseUp' bitmask on rk to GEM_MB_ONGOING_ACTION

This way, we can identify `MouseUp' when unintended by modified
`rk_flags'.

MarcelHB committed:
GUI::*: rewrites bit pattern check in `OnMouseUp's

MarcelHB committed:
IWD2: enables dark main menu bg when playing at night

In the original game, playing between 6:00pm and 6:59am yields a
different background with a flare animation. Cozy.

MarcelHB committed:
GUI::Window: adds support for key stroke handling

If we intend to support certain key strokes in an abstract window-scope,
we need support for callbacks.

MarcelHB committed:
GUI::EventMgr: polls current window for pressed key use

But only if the control is not doing anything.

MarcelHB committed:
GUIScript::PyHelpers: refactors code, adds window specialization

MarcelHB committed:
GUIScript: adds `_GemRB.Window_SetKeyPressEvent' method

For Py scripts to set their own window-scoped callbacks.

MarcelHB committed:
PythonHelpers: fixes compatibility with GCC

Otherwise, it yields an error for missing declarations.

kujeger committed:
SDL2: Fix RMB stuck cursor

If not using touch input,
SDL_MOUSEMOTION/SDL_MOUSEBUTTONDOWN/SDL_MOUSEBUTTONUP would never break,
instead moving on to SDL_KEYDOWN -- which would read
event.key.keysym.scancode and erronously break on the scancode event
read when RMB+MouseMovement occurred, never moving on to return the
event to SDLVideoDriver::ProcessEvent . We now explicitly return the
event when event.button.which is not SDL_TOUCH_MOUSEID .

kujeger committed:
SDL2: remove unreachable break statement

bradallred committed:
Merge pull request #51 from kujeger/sdl2-mouse

SDL2: Fix RMB stuck cursor

MarcelHB committed:
IWD2: updates some avatar size values

Value range taken from BG2. Otherwise, lots of (N)PCs lack equipment
animations.

bradallred committed:
Merge pull request #52 from MarcelHB/iwd2_avatar_sizes

IWD2: updates some avatar size values

lynxlynxlynx committed:
iwd2: externalised the racial bonuses to skill points

lynxlynxlynx committed:
iwd2: always set IE_UNUSED_SKILLPTS, so old values don't carry over

lynxlynxlynx committed:
CloseContainerWindow: resolved iwd2 fixme

lynxlynxlynx committed:
iwd2: removed unused guiworld var

a leftover from initial gui work - all other games use it

lynxlynxlynx committed:
iwd: removed useless todo and fixme

lynxlynxlynx committed:
iwd: removed some useless commented out code

remnants of the transition to immediate saving of stats

lynxlynxlynx committed:
iwd2: another skills change to be clearer and work properly for squatters

lynxlynxlynx committed:
Merge pull request #50 from MarcelHB/window_keys

GUIScript: allows callbacks for window-scoped key presses

MarcelHB committed:
Merge remote-tracking branch 'gemrb/master' into recent_master

MarcelHB committed:
Core/GUIScript/SDL: post-merge fixes

bradallred commented:
oh wow! Just about to hit the sack here. I'll have to digest this over lunch tomorrow. Looks like a lot of good work, so thank you :)

Edit: is this up-to-date with the main master now?

MarcelHB commented:
It's as up-to-date as can be. I was not really sure how to fully solve some conflicting things around SDL but since your are in WIP there as well, I guess this will need further work anyway.

bradallred commented:
ah, ok. I imagine you are referring to SDL2 conflicts; I don't recall any SDL changes outside that. I believe all those conflicts can simply be discarded (most changes there were complete hacks or addressed elsewhere now).

Were you able to build with SDL 1.2 for now, or should I prioritize getting the SDL 2 driver reassembled for you?

bradallred commented:
I believe that `GemRB_Window_SetPicture` is simply replaced with `GemRB_View_SetBackground`.

If you require additional customization for the `Window` version of that; doing that on the Python side would be better then having a GUIScript method for it. There are several examples of this in `GUIClasses.py`

MarcelHB commented:
Ok, good to know.

SDL 1.2 works well, as long as we talk about my Linux desktop VM.

bradallred commented on gemrb/core/GUI/EventMgr.h:
```diff
@@ -74,6 +74,8 @@ class Window;
 #define GEM_MB_SCRLUP           8
 #define GEM_MB_SCRLDOWN         16
 
+#define GEM_MB_ONGOING_ACTION   33
```
This looks unused? Also the new Event struct has a repeat count field and possibly other useful attributes that could/should be used instead I think.

If it would be helpful, we can alter the `OnKeyX` and `OnMouseX` methods to directly take the event object so all that stuff is available there. I've already (locally) changed the key handlers to take the `KeyboardEvent` directly, but it would be no big thing to use the raw event instead; thats what most GUI libraries seem to do anyway.

bradallred commented on gemrb/plugins/GUIScript/GUIScript.cpp:
```diff
@@ -1638,6 +1638,48 @@ static PyObject* GemRB_Control_SetText(PyObject* self, PyObject* args)
 	Py_RETURN_NONE;
 }
 
+PyDoc_STRVAR( GemRB_Window_SetKeyPressEvent__doc,
```
this is at the very least going to need a rewrite since we don't have a "window index" anymore. nor does `core` have APIs for getting at the windows.

Can't this be handled by `Window::RegisterHotKeyCallback`? We have a Python variant for `EventCallback`

If you are just trying to use this with buttons then `button.SetHotkey` should do what is needed already.

MarcelHB replied:
I've updated the doc, but it is already rewritten internally. If not needed anymore, then we can also purge it.

bradallred commented on gemrb/plugins/GUIScript/PythonHelpers.cpp:
```diff
@@ -111,3 +75,62 @@ void PythonControlCallback::operator() (Control* ctrl) const
 }
+
+template <>
+bool PythonObjectCallback<WindowKeyPress>::operator() (WindowKeyPress *wkp) {
```
This too ought to already be handled by the existing python callbacks. The python representation of the window (`GWindow`) would be passed as the parameter to the python function.

bradallred commented:
Most of the (C++) Python changes look good (left a note about one), but I'm mostly trusting that it works :p 
this being a WIP branch I'm not concerned about breaking enough to thoroughly vet it.

MarcelHB commented:
Ok, I hope I can go through your comments tomorrow.

bradallred commented:
So looking between this and the pull requests from the main repo, it looks like all of the Python and window event changes are only there to support activating PCs with the number keys. The equivalent on this branch would normally simply be to call `SetHotkey` on the buttons in the same loop that is doing the portrait window setup...

Pretty sure that won't work as is, however, since it probably just set the hotkey for when the button's window has focus. What we _could_ do here is have `GameControl` setup a callback with `EventMgr:: RegisterEventMonitor`, but that has an unwanted side-effect of changing selected PC even when just trying to type a number in an input field...

Perhaps the best approach would be to funnel _unhandled_ key events to the `WindowManager:: gameWin` which would end up being delivered to `GameControl` and that would work, but then GameControl could receive events while a modal window is displaying... which we actually would want for the PC selection, but nothing else (that I know of). So the callback would have to check `WindowManager:: IsPresentingModalWindow` and bail for non-numeric keys.

That might have a nice sideeffect of being able to have all the hot key logic for both the ActionBar and PortraitWindow in a single concise location (inside `GameControl`) rather then sprinkled about in the Python scripts (likely duplicated all over the place) with funky one-off GUIScript APIs

What do you think? I'll play around tonight and see if I can get something pushed up to discuss.

bradallred commented:
so... it turns out that apparently I already thought of this and it actually works right now (I had to push a one line change to ensure the `GameControl` is the focus of its window).

The problem is that the window's python "update script" needs to get triggered...

It would probably be cleanest to give `View` a refresh event handler that we could call everytime the `MarkDirty` is called (probably have it check the existing value of the dirty flag to avoid extra calls). This would also be much less manual then we currently do too... I'm pretty sure there are several places where we assign an event call back just so we can call the update function. Might fix a few bugs too (like char hp not refreshing while looking in the inventory etc)

bradallred commented:
nevermind 😀 

turns out the failure to refresh was a missing `SelectPCSingle` call. It was actually your pull request on the main repo that made me realize that! so thank you.

tl;dr
Using the number buttons to make PC selections seems to work fine here now.

MarcelHB committed:
GUIScripts::GUIClasses: fixes odd method signature

In `CreateButton'. Things start to work again.

MarcelHB committed:
GUI::EventMgr: removes outdated macro definition

MarcelHB committed:
GUIScript: updates doc for `Window_SetKeyPressEvent'

MarcelHB committed:
GUIScripts::IWD2: brings main menu back to work (at night)

As `SetPicture' has been replaced by `SetBackground'.

MarcelHB commented:
Ok, I have taken care of the minor things right now.

I have already spotted some weird displacements of button windows in BG, but I suggest to take care of that in the next steps.

bradallred commented:
I set up Travis today to help fix the compile warnings. I probably should find the time to fix the 100 or so precision loss and dead code warnings (a few others too) in the main repo so that I can treat warnings as errors again on my local machine.

I haven't noticed any "weird displacements", but perhaps I simply need to run with a different resolution (I mainly run 640x480 so I can run and debug easily on my laptop). I will take a look when I get home.

I think the final question to answer before I merge this is: what is the purpose of `Window_SetKeyPressEvent` and `WindowKeyPressHandler`? They seem superfluous in this branch...

MarcelHB commented:
Going to test this tonight, otherwise I will remove it.

bradallred commented:
Can you enlighten me as to what 4758aa3 was specifically supposed to fix? there isn't much to go on in the commit message.

Apparently that conflicted with changes I had made and now with new changes I just pushed up. Was this tested? I'm concerned your merge strategy was to select your changes then fixup the compile errors and I'm uncertain if anything important of mine got dropped or misplaced 😛 

Like I said, not too concerned with breaking anything on this branch, I'd just rather know what the intent was and if you tested it (not that everything was working 100% there anyway).

MarcelHB commented:
Sorry, was a little out of time.

I just introduced a new event-bit there to have some distinction between the physical `OnMouseUp` events and that fake one to make pressed keys work as if e.g. buttons were pressed again. This was necessary to make the scrollbar work in the CG of IWD2 after enabling repeating keys as it lost focus immediately, rendering dragging impossible.

IWD2 was back online, but I didn't test that specific fix in your branch yet.

So the answer is: I just rewrote the bit-pattern checks to remain compatible.

bradallred commented:
I have/will be removing all of our phony events in this branch so that wont be necessary (and likely doesn't work here now). It never was a good way of doing things, it was just an easy shortcut we used because of existing poor design.

The repeat event stuff will be moved directly into the `Control` class eventually. Really just need to refactor some of the event handlers into a common "action" handler. Once that happens a mouse down event can start a timer that repeatedly fires the controls action callback and a mouse up will cancel that timer. Much less hacky and will lend itself to cleaner more modular code rather then our tangled event handling methods we have now.

The focus problems are a separate issue, but those too should be addressed here in a different way then we have historically, if it isn't already.

The entire scrollbar -> TA or Window relationship was scrapped and that behavior is being rewritten in the `ScrollView` class. `Scrollbar` only needs to be a dumb class that reflects the state of something else (usually a `ScrollView`) and fires a callback when it changes its value so the lister can update its state rather then the scrollbar actually knowing about and doing that directly.

I've been pushing up bits and pieces of that here and there, but I'm still working on some details that aren't quite ready yet.

MarcelHB commented:
Ah, I remember ... IWD2 load screen gets stuck by some Python exceptions. Gonna take a look into them.

MarcelHB commented:
By the way, this is my BG on this branch right now.
![screenshot_bg](https://cloud.githubusercontent.com/assets/238558/15453320/bfa82dc4-2011-11e6-8892-561d2e65eb7d.jpg)

MarcelHB committed:
GUIScripts::IWD2::LoadScreen: changes selection of background

I could not find `GUILxx' anywhere, and since `View_SetBackground'
discards the attempt to load non-existing pictures with an exception, we
should no longer keep this.

The load screen behaviour looks unchanged anyway so far.

MarcelHB commented:
Same on your `subviews` branch, ScollBars (IWD2) work nowhere as well so far.

In the scripts, lots of things seem to be broken around the dropped `LoadWindowPack` and my attempts to play with `LoadWindow` instead only caused more problems deeper down the stack, such as `GetControl`s failing to get anything.

And I don't get the Travis error. Is it some static analysis setting? Here, everything looks good, for evident reasons.

MarcelHB committed:
Merge branch 'brad/subviews' into recent_master: solves conflict

bradallred commented:
Right, I thought I mentioned in my email that _no other games_ besides BG2 have had their GUIScripts updated (and even then BG2 still has some left to do). That is why you see what you see in BG and why IWD2 is throwing exceptions. Those windows clustered in the center just need to have their position set like their counterparts in `GUIScripts/bg2/MessageWindow.py`; look for `WINDOW_LEFT` and such. It looks like other problems are occurring as well tho.

Yes, `LoadWindowPack` is gone and replaced with `LoadWindow`. If you compare the BG2 counterpart scripts you should see that you can delete _a lot_ of code in the `OpenXWindow` functions and replace it with a few lines. Look for `GUICommonWindows.OpenTopWindow`.

The window management on the Python side still needs some work. I want to basically make a window manager that we can just push windows onto and when we close them it should automatically refresh the new top window automatically. right now all of the subwindows are completely done manually with lots of global variables tracking the windows; real messy stuff. We should be able to rid ourselves of that easily now. This is probably my next area of focus along with the `ScrollView`

Re scrollbars: they have been disconnected right now because I am working on the `ScrollView` class. There is no longer an easy way to reconnect them since I have now removed the `Scrollbar` member from `View`. It will be a while before they work again :(

There should be plenty that you can do for IWD2 or whatever other game without scrollbars I would think :) Really just getting it to the point where the game loads up correctly would be a godsend and 95+% of the work unique to IWD/IWD2/BG; PST may need a bit more love, I don't remember the status of its little floating window menu.

The Travis Error is my fault: `error: assigning to 'unsigned short' from incompatible type 'const GemRB::KeyboardEvent' wkp.key = key;`

The Travis wont work on my branch until this is merged in because I'm still missing some Travis stuff from main master.

This pull request is getting too long as it is. I'm just going to merge it in and revert the stuff that looked unneeded for now. We can pull those bits back in if we need them later on.

Thanks again for all your help with this 😃

MarcelHB commented:
:+1:

Will go through IWD2/BG over time and find things to fix that will hopefully not interfere with your work.
